### PR TITLE
pkg/nodeaddress: give preference to k8s IP allocation

### PR DESCRIPTION
**Pull request by aanm**
When using cilium with k8s, we should prioritize k8s pod CIDR for the
node CIDR allocation. Otherwise cilium might use the same PodCIDR on 2
different nodes, such as:

(The IPv6 pod cidr was automatically derived from the 1st global scope
interface where on this case the global scope interface had the same IP
private IP address)

```
$ kubectl get nodes -o json | grep -i CIDR
"io.cilium.network.ipv4-pod-cidr": "10.0.0.0/16",
"io.cilium.network.ipv6-pod-cidr": "f00d::a00:20f:0:0/96",
"podCIDR": "10.0.0.0/16"
"io.cilium.network.ipv4-pod-cidr": "10.1.0.0/16",
"io.cilium.network.ipv6-pod-cidr": "f00d::a00:20f:0:0/96",
"podCIDR": "10.1.0.0/16"
```

Or in case the podCIDR was an IPv6 address:

```
$ kubectl get nodes -o json | grep -i CIDR
"io.cilium.network.ipv4-pod-cidr": "10.15.0.0/16",
"io.cilium.network.ipv6-pod-cidr": "fd02::/96",
"podCIDR": "fd02::/96"
"io.cilium.network.ipv4-pod-cidr": "10.15.0.0/16",
"io.cilium.network.ipv6-pod-cidr": "fd02::1:0:0/96",
"podCIDR": "fd02::1:0:0/96"
```

This commit derives the IPv6 pod CIDR from the IPv4 pod CIDR and
vice-versa and gives priority to the podCIDR set by kubernetes if the
user runs cilium in non-overlay mode.

Signed-off-by: André Martins <andre@cilium.io>

### CODE REVIEW

File: daemon/daemon.go
Comment (ianvernon) [inline daemon/daemon.go]: Lines 888-889 vs. 891-892 seem redundant to me. Remove one of the two sets of lines.

File: pkg/nodeaddress/node_address.go
Comment (ianvernon) [inline pkg/nodeaddress/node_address.go]: ... then the IPv4 allocation should be derived from the IPv6AllocRange.

File: pkg/nodeaddress/node_address.go
Comment (ianvernon) [inline pkg/nodeaddress/node_address.go]: add --> added

File: pkg/nodeaddress/node_address.go
Comment (ianvernon) [inline pkg/nodeaddress/node_address.go]: The IPv6 allocation should be derived from the IPv4 allocation.

File: pkg/nodeaddress/node_address.go
Comment (ianvernon) [inline pkg/nodeaddress/node_address.go]: add --> added

**Review by ianvernon — APPROVED** [review]
Logic seems sound, I just have some log message / comment grammar nitpicks. Feel free to fix my comments and merge after you fix them without asking for a second review.

File: pkg/nodeaddress/node_address.go
Comment (aanm) [reply pkg/nodeaddress/node_address.go]: Fixed

File: pkg/nodeaddress/node_address.go
Comment (aanm) [reply pkg/nodeaddress/node_address.go]: Fixed

File: pkg/nodeaddress/node_address.go
Comment (aanm) [reply pkg/nodeaddress/node_address.go]: Fixed

File: pkg/nodeaddress/node_address.go
Comment (aanm) [reply pkg/nodeaddress/node_address.go]: Fixed

File: daemon/daemon.go
Comment (aanm) [reply daemon/daemon.go]: I've reordered the sentences

**Review by nebril — APPROVED** [review]
LGTM.

File: pkg/nodeaddress/node_address.go
Comment (ianvernon) [inline pkg/nodeaddress/node_address.go:134]: Good example, thank you for adding this!

**Review by ianvernon — APPROVED** [review]
LGTM!

**Pull request merged**
